aws/rust-runtime/aws-types/src/profile/parse.rs

```diff
@@ -0,0 +1,333 @@
+/// Profile parsing is actually quite strict about what is and is not whitespace, so use this instead
+/// of `.is_whitespace()` / `.trim()`
+const WHITESPACE: &[char] = &[' ', '\t'];
+const COMMENT: &[char] = &['#', ';'];
+
+/// Location for use during error reporting
+#[derive(Clone, Debug, Eq, PartialEq)]
+struct Location {
+    line_number: usize,
+    path: String,
+}
+
+/// An error encountered while parsing a profile
+pub struct ProfileParseError {
+    /// Location where this error occurred
+    location: Location,
+
+    /// Error message
+    message: String,
+}
+
+impl Display for ProfileParseError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "error parsing {} on line {}:\n  {}",
+            self.location.path, self.location.line_number, self.message
+        )
+    }
+}
+
+/// Validate that a line represents a valid subproperty
+///
+/// - Subproperties looks like regular properties (`k=v`) that are nested within an existing property.
+/// - Subproperties must be validated for compatibility with other SDKs, but they are not actually
+/// parsed into structured data.
+fn validate_subproperty(value: &str, location: Location) -> Result<(), ProfileParseError> {
+    if value.trim_matches(WHITESPACE).is_empty() {
+        Ok(())
+    } else {
+        parse_property_line(value)
+            .map_err(|err| err.into_error("sub-property", location))
+            .map(|_| ())
+    }
+}
+
+fn is_empty_line(line: &str) -> bool {
+    line.trim_matches(WHITESPACE).is_empty()
+}
+
+fn is_comment_line(line: &str) -> bool {
+    line.starts_with(COMMENT)
+}
+
+/// Parser for profile files
+struct Parser<'a> {
+    /// In-progress profile representation
+    data: RawProfileSet<'a>,
+
+    /// Parser state
+    state: State<'a>,
+
+    /// Parser source location
+    ///
+    /// Location is tracked to facilitate error reporting
+    location: Location,
+}
+
+enum State<'a> {
+    Starting,
+    ReadingProfile {
+        profile: &'a str,
+        property: Option<&'a str>,
+        is_subproperty: bool,
+    },
+}
+
+/// Parse `file` into a `RawProfileSet`
+pub fn parse_profile_file(file: &File) -> Result<RawProfileSet, ProfileParseError> {
+    let mut parser = Parser {
+        data: HashMap::new(),
+        state: State::Starting,
+        location: Location {
+            line_number: 0,
+            path: (&file.path).to_string(),
+        },
+    };
+    parser.parse_profile(&file.contents)?;
+    Ok(parser.data)
+}
+
+impl<'a> Parser<'a> {
+    /// Parse `file` containing profile data into `self.data`.
+    fn parse_profile(&mut self, file: &'a str) -> Result<(), ProfileParseError> {
+        for (line_number, line) in file.lines().enumerate() {
+            self.location.line_number = line_number + 1; // store a 1-indexed line number
+            if is_empty_line(line) || is_comment_line(line) {
+                continue;
+            }
+            if line.starts_with('[') {
```

Comment:
Is whitespace before the `[` valid?

Reply:
I don't think so—this code was ported directly from the reference implementation but I will confirm with the SEP

Reply:
confirmed, whitespace isn't valid there